BUG FIX: 'dump_logs_by_extension out' would also dump *.Rout files

HenrikBengtsson opened the pull request:
I ended up getting ["This log is too long to be displayed. Please reduce the verbosity of your build or download the raw log."](https://travis-ci.org/HenrikBengtsson/matrixStats/builds/47803485) (see end of page) from Travis.  This was because `dump_logs_by_extension out` would also dump *.Rout files, e.g.

```
$ find ./matrixStats.Rcheck -type f -name '*out'
./matrixStats.Rcheck/00install.out
./matrixStats.Rcheck/matrixStats-Ex.Rout
./matrixStats.Rcheck/tests/allocArray.Rout
[...]
```
Note that this would also cause `dump_logs` to output all *.Rout files.

gaborcsardi commented:
Thanks. Can't you just specify `.out` as extension? With the dot included? I feel that putting the dot in build the script restricts the files that can be dumped this way.

HenrikBengtsson commented:
Sure I can, and I also first did that. But, at least to me, technically the dot is _not_ part of the file name extension, i.e. a misleading function name.  Also, coming from R and grep etc., it was not clear whether one would need to use `\.out`, `[.]out` or just `.out`.  I would still argue that one should _not_ specify the dot.

Also, if dots should be specified, then the following needs to be updated to:
```
DumpLogs() {
    echo "Dumping test execution logs."
    DumpLogsByExtension ".out"
    DumpLogsByExtension ".Rout"  ## ?!?
    DumpLogsByExtension ".log"
    DumpLogsByExtension ".fail"
}
```

So, I favor not to have to specify the dot.  It's clearer and less of a hack.

For full power, one could add (untested):

```
DumpByPattern() {
    if [[ -z "$1" ]]; then
        >&2 echo "dump_by_pattern requires exactly one argument, got: $@"
        exit 1
    fi
    pattern=$1
    shift
    package=$(find . -maxdepth 1 -name "*.Rcheck" -type d)
    if [[ ${#package[@]} -ne 1 ]]; then
        >&2 echo "Could not find package Rcheck directory, skipping dump."
        exit 0
    fi
    for name in $(find "${package}" -type f -name "${pattern}"); do
        echo ">>> Filename: ${name} <<<"
        cat ${name}
    done
}
```
and then I guess:
```
DumpLogsByExtension() {
    if [[ -z "$1" ]]; then
        >&2 echo "dump_logs_by_extension requires exactly one argument, got: $@"
        exit 1
    fi
    DumpByPattern "*.$1"
}
```

HenrikBengtsson commented:
I've updated this PR with a local DumpByPattern() function and added 'dump_by_pattern' to the external API.  This can be useful for troubleshooting/dumping for instance a specific *.tex file generated during vignette processing.

I've verified it on Travis CI.

For full power, one could consider using `-wholename` instead of `-name` for the Unix `find` command.  That would allow to filter by pathname (=path + filename) rather that just filename.

gaborcsardi commented:
This is good, I only have one concern, about compatibility with `r-travis`. But it probably does not affect many people (not that they are switching to `r-builder`, anyway. :)). So, thanks!

HenrikBengtsson commented:
Thanks. Let's make r-builder clearly better than r-travis and people will follow :)

gaborcsardi commented:
I'd prefer merging it back to r-travis, actually. But of course I don't mind making it better, either. :)

HenrikBengtsson commented:
Sure.

Also, if it's not clear, you've got my permission to do whatever you want with my contributions here.  I consider them to be contributed under whatever open-source license you use here.  ...unless Microsoft's buys you, then I want my share ;)

gaborcsardi commented:
Well, but if Microsoft buys me, this will not be funny (for me). :) So, yes, I need a license, and also a CLA.... :)
